Begin using C++ coroutines for IPC messages without a reply

achristensen07 opened the pull request:
#### 655b6c1d090a5812e07d991346dbf21a04142e4f
<pre>
Begin using C++ coroutines for IPC messages without a reply
<a href="https://bugs.webkit.org/show_bug.cgi?id=292067">https://bugs.webkit.org/show_bug.cgi?id=292067</a>
<a href="https://rdar.apple.com/150042541">rdar://150042541</a>

Reviewed by Richard Robinson.

IPC receivers without a reply don&apos;t have anything to &quot;return&quot; to the
sending process, so we only need to check for Awaitable&lt;void&gt; as a return
type.  This syntax makes it so we don&apos;t need to explicitly capture each
variable for each async step.

* Source/WebKit/Platform/IPC/HandleMessage.h:
(IPC::callMemberFunctionCoroutine):
(IPC::MethodSignatureValidation&lt;R):
(IPC::handleMessage):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::convertRectToMainFrameCoordinates):
(WebKit::WebPageProxy::showCorrectionPanel):
* Source/WebKit/UIProcess/WebPageProxy.h:

Canonical link: <a href="https://commits.webkit.org/294160@main">https://commits.webkit.org/294160@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e7c8077a241d4d4f68c31d30a4d1a48baeaa122e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100882 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20547 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10847 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/106027 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51482 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20854 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/29036 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76824 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33864 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103889 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16040 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91130 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57170 "Found 1 new API test failure: /WPE/TestCookieManager:/webkit/WebKitCookieManager/replace-get-all-cookies (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15854 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50856 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/85769 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9213 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108379 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28005 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/20601 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/85789 "Passed tests") | 
| [⏳ 🧪 services ](https://ews-build.webkit.org/#/builders/Services-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28373 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87330 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85327 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30049 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7779 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/22038 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16427 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27938 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/33212 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27749 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31072 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29312 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -12851,6 +12851,13 @@ void WebPageProxy::convertRectToMainFrameCoordinates(WebCore::FloatRect rect, st
 
+Awaitable<std::optional<WebCore::FloatRect>> WebPageProxy::convertRectToMainFrameCoordinates(WebCore::FloatRect rect, std::optional<WebCore::FrameIdentifier> frameID)
+{
+    co_return co_await AwaitableFromCompletionHandler<std::optional<WebCore::FloatRect>> { [protectedThis = Ref { *this }, rect, frameID] (auto completionHandler) {
```
I wonder if there's any template deduction guide that can be written to make it so that the template type can be inferred in this line

achristensen07 replied:
I don't think so because the argument is a lambda with an auto parameter.  The type needs to be somewhere.

rr-codes commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -3075,7 +3076,7 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
 #if PLATFORM(MAC)
     void substitutionsPanelIsShowing(CompletionHandler<void(bool)>&&);
-    void showCorrectionPanel(WebCore::AlternativeTextType panelType, const WebCore::FloatRect& boundingBoxOfReplacedString, String&& replacedString, String&& replacementString, Vector<String>&& alternativeReplacementStrings, WebCore::FrameIdentifier);
+    Awaitable<void> showCorrectionPanel(WebCore::AlternativeTextType panelType, WebCore::FloatRect boundingBoxOfReplacedString, String replacedString, String replacementString, Vector<String> alternativeReplacementStrings, WebCore::FrameIdentifier);
```
why remove the r-values out of curiosity?

achristensen07 replied:
references go out of scope when the first co_await makes the coroutine return control (unless the reference is copied or moved onto the call frame), but parameters passed by reference have the same lifetime as the coroutine.

achristensen07 committed:
Begin using C++ coroutines for IPC messages without a reply
https://bugs.webkit.org/show_bug.cgi?id=292067
rdar://150042541

Reviewed by Richard Robinson.

IPC receivers without a reply don't have anything to "return" to the
sending process, so we only need to check for Awaitable<void> as a return
type.  This syntax makes it so we don't need to explicitly capture each
variable for each async step.

* Source/WebKit/Platform/IPC/HandleMessage.h:
(IPC::callMemberFunctionCoroutine):
(IPC::MethodSignatureValidation<R):
(IPC::handleMessage):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::convertRectToMainFrameCoordinates):
(WebKit::WebPageProxy::showCorrectionPanel):
* Source/WebKit/UIProcess/WebPageProxy.h:

Canonical link: https://commits.webkit.org/294160@main
